Feat: add modern skin for smapp and separate themes to own configuration

maparr opened the pull request:
It closes #889 
## Apply skin integration from settings tab
<img width="1181" alt="Screenshot 2022-05-18 at 23 57 51" src="https://user-images.githubusercontent.com/54288612/169168716-0e87ca21-9554-41f7-b7a1-5fbb8ba14a98.png">


## How it works
Depending on the user preferences, the client can select new skin, buy the default modern skin is the default

maparr committed:
fix: [skip ci] separate themes
feat: [skip ci] add modern theme

maparr committed:
feat: add modern skin and apply fixes for other themes

maparr committed:
Merge branch 'develop' into feature/modern-skin

brusherru commented on app/theme/variants/light.ts:
```diff
@@ -0,0 +1,142 @@
+import colors from '../colors';
+import { backup, tooltip } from '../../assets/images';
+
+export default {
```
I think we need a common interface for the variants to help us not to miss something 🤔

maparr replied:
I added a [module declaration](https://github.com/spacemeshos/smapp/pull/908/files#diff-7476b139c90076813fee5a70dd7f7d878501a338278fc9aa5a703dea2c63d6f3R7) for our skins,  if we changing the skins file, just need to add the same by meaning to other files. Theme autocomplete working for styled-components

brusherru commented on app/redux/ui/actions.ts:
```diff
@@ -14,7 +15,10 @@ export const setOsTheme = () => async (dispatch: AppThDispatch) => {
 };
 
-export const switchTheme = () => ({ type: THEME_SWITCHER });
+export const switchTheme = (type: number) => ({
```
Seems we need to store the last value of it in a persistent store to use this value on the next app session.

Only renderer requires this piece of data, so I think you can just use `localStorage` for now as the easiest solution.
If we will need this data in the main process we can easily switch to storing it on fs (via IPC and `StoreService`).

Mb write a tiny middleware for it :)

maparr replied:
yes, it just was a working solution for testing, replacing this solution with local storage.

brusherru reviewed (commented):
Left just a couple of comments.

Just in case, check the dropdown in modern skin on SendTransaction screen:
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/1897530/169212552-263a7fe9-f641-43e6-bf25-e8b07cc2f4c4.png">

Waiting for the PR that will be ready for review & tests.
Hopefully something will happen with all the corners :D

maparr committed:
fix: [skip ci] eslint/prettier

maparr committed:
fix: [skip ci] replace design for Autocomplete list dropdown and Amount field on the TxPrams.tsx

maparr committed:
chore: [skip ci] remove unused Readme for themes

maparr committed:
feature: [skip ci] selecte modern skin for new clients and based on OS for clients with existing wallet if skin not selected

maparr committed:
fix: [skip ci]
- fix address filed to match skins
- not render list if no addresses

maparr committed:
fix:
- disable state for amount field
- add left side panel svg for dark mode
- clean up not used icons for modern design
- delete custom styles for sub dropdown items

maparr committed:
fix: eslint
- sort out Settings.tsx
- remove isDarkMode for tooltip ( using from theme )

maparr committed:
fix: update skins and replace disable state for Input to normal state

maparr committed:
fix: update select a wallet ( fixed height )

maparr committed:
fix:
- add border radius to check icon/indicator
- revert Commitment component for POS
- add border radius to all boxes in the app

avive commented:
Great job guys - playing with the build and went over it with @oriya this morn.

maparr committed:
fix: add space to the small button text padding

maparr committed:
fix: [skip ci] add chevron for modern theme and add radius for numeric indicators

maparr committed:
fix:
- simplify fonts for themes
- add universal lts td font for modern dark theme

maparr committed:
fix:
- simplify fonts for themes
- add universal lts td font for modern dark theme

maparr committed:
fix: revert new font

maparr committed:
feat: simplify skin id detection

maparr committed:
fix: add contacts , remove the border for modern skin

brusherru commented on app/components/transactions/RewardRow.tsx:
```diff
@@ -64,7 +64,7 @@ const BlackText = styled(Text)`
 `;
 
 const BoldText = styled(Text)`
```
It would be perfect to extract such styled components into "basicComponents" and then reuse them in other places.
It will reduce the codebase, and make code more maintainable and clear.
In case we want to merge it ASAP — we can postpone this tweak.

brusherru replied:
See also HeaderText, SubHeaderText, etc
We have a lot of code duplication of styled-components.

maparr replied:
no, I think I know how to add it in a short way. 
For **dropdown lists**, I think we have to create a feature task that will combine it with different styles, but everything will be in the dropdown component. I won't do it right now because a lot of moments have to be covered for this, it just will take this PR to a different category. 
For other components, I think step by step let's highlight it through the feature task and refactor it separately , to be sure PRs will be small

brusherru commented on app/theme/index.ts:
```diff
@@ -0,0 +1,53 @@
+export const getClientSettingsTheme = (): string | null =>
+  window.localStorage.getItem(CLIENT_SETTINGS_THEME_KEY);
+
+export const SKINS = {
```
Why not enum?

brusherru replied:
Then you can add
```
const THEMES = {
  [Skins.Modern]: modern,
  [Skins.Dark]: dark,
  [Skins.Light]: light,
};
```

and tweak a bit

```
// rename proposal for isValidSettingsTheme
export const isValidThemeId = (id: string): boolean =>
  Object.keys(THEMES).includes(id);

// rename proposal
export const getThemeById = (skin: Skins | null, isDark: boolean) => {
  if (!skin) return isDark ? dark : light;
  return THEMES[skin];
};
```

brusherru commented on app/theme/index.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+};
+
+export type DefaultSkin = typeof modern | typeof light | typeof dark;
```
Unused?
And a confusing name.
Do you mean `AnyTheme`? :)

maparr replied:
It used, [here](https://github.com/spacemeshos/smapp/pull/908/files#diff-7476b139c90076813fee5a70dd7f7d878501a338278fc9aa5a703dea2c63d6f3R7), to make these themes configuration and styled-components props declared, it's from the styled component article like [this](https://medium.com/rbi-tech/theme-with-styled-components-and-typescript-209244ec15a3). Default Theme is declared type in a styled component that I'm rewriting.
naming will be changed.

maparr replied:
found why is not using, removed.

brusherru reviewed (changes requested):
Left a few comments.

Also, something happened with "send tx" screen:
- the address input width is smaller until it turns into a dropdown
- btw, should not it also search by a nickname? 🤔  probably we have to open an issue for this tweak
- Fee dropdown: arrow moved to the left a bit

Check out screenshots:
<img width="534" alt="image" src="https://user-images.githubusercontent.com/1897530/170947734-017af138-ef67-4f4b-a3db-e3efe84bf465.png">
<img width="534" alt="image" src="https://user-images.githubusercontent.com/1897530/170947793-9e1e6d95-3a7e-4cef-adfe-0e27f9a8f2ff.png">

maparr commented:
> Left a few comments.
> 
> Also, something happened with "send tx" screen:
> 
> * the address input width is smaller until it turns into a dropdown
> * btw, should not it also search by a nickname? 🤔  probably we have to open an issue for this tweak
> * Fee dropdown: arrow moved to the left a bit
> 
> Check out screenshots: <img alt="image" width="534" src="https://user-images.githubusercontent.com/1897530/170947734-017af138-ef67-4f4b-a3db-e3efe84bf465.png"> <img alt="image" width="534" src="https://user-images.githubusercontent.com/1897530/170947793-9e1e6d95-3a7e-4cef-adfe-0e27f9a8f2ff.png">

yeah, about this, so by default we don't have any items in the dropdown that looked very wired, so I switched, if we don't have any items in the auto-complete component, it will be just an input. If we have some data, we will show a dropdown.
cc: @avive , @oriy  what do you think about this changed flow.

avive commented:
- The auto-complete should def accept an text input and complete to any contacts nickname
- All nicknames should be displayed in the drop-down
- Ideally, auto-complete should also include 5 most recent addresses from last previous transactions
- It is okay not to display the drop-down if there are 0 items to show

maparr commented:
> * The auto-complete should def accept an text input and complete to any contacts nickname
> * All nicknames should be displayed in the drop-down
> * Ideally, auto-complete should also include 5 most recent addresses from last previous transactions
> * It is okay not to display the drop-down if there are 0 items to show

yes, seems we have to add predefined contacts that the client added, and automatically the auto-complete will be having the dropdown view. I will create a feature and issue task for these moments. Right now it's not a blocking issue. cc: @brusherru

maparr committed:
fix:
- add bold text as separate component
- simplify theme detection

maparr committed:
fix: transactions light screen

brusherru reviewed (approved):
LGTM
